zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -521,9 +460,111 @@ impl Chain {
         self.sprout_trees_by_height.get(&height).cloned()
     }
 
-    /// Returns the Sapling
-    /// [`NoteCommitmentTree`](sapling::tree::NoteCommitmentTree) specified by a
-    /// [`HashOrHeight`], if it exists in the non-finalized [`Chain`].
+    /// Add the Sprout `tree` to the tree and anchor indexes at `height`.
+    ///
+    /// `height` can be either:
+    /// - the height of a new block that has just been added to the chain tip, or
+    /// - the finalized tip height: the height of the parent of the first block of a new chain.
+    fn add_sprout_tree_and_anchor(
+        &mut self,
+        height: Height,
+        tree: Arc<sprout::tree::NoteCommitmentTree>,
+    ) {
+        // Having updated all the note commitment trees and nullifier sets in
+        // this block, the roots of the note commitment trees as of the last
+        // transaction are the anchor treestates of this block.
+        //
+        // Use the previously cached root which was calculated in parallel.
+        let anchor = tree.root();
+        trace!(?height, ?anchor, "adding sprout tree");
+
+        // TODO: fix test code that incorrectly overwrites trees
+        #[cfg(not(test))]
+        assert_eq!(
+            self.sprout_trees_by_height.insert(height, tree.clone()),
+            None,
+            "incorrect overwrite of sprout tree: trees must be reverted then inserted",
+        );
+        #[cfg(not(test))]
+        assert_eq!(
+            self.sprout_anchors_by_height.insert(height, anchor),
+            None,
+            "incorrect overwrite of sprout anchor: anchors must be reverted then inserted",
+        );
```

Comment:
Optional/Nitpick: group conditionally compiled statements in blocks?

```suggestion
        {
          assert_eq!(
              self.sprout_trees_by_height.insert(height, tree.clone()),
              None,
              "incorrect overwrite of sprout tree: trees must be reverted then inserted",
          );
          assert_eq!(
              self.sprout_anchors_by_height.insert(height, anchor),
              None,
              "incorrect overwrite of sprout anchor: anchors must be reverted then inserted",
          );
        }
```

Reply:
Changed in 6a26fe7ba7